torch/_inductor/codegen/common.py

```diff
@@ -1789,13 +1814,17 @@ def generate(
                         line = f"{expr}{self.suffix}"
                     buffer.writeline(line)
 
+                    # cpp backend cannot determin is_vec at this point
                     if (
                         assignment
-                        and config.test_configs.runtime_triton_dtype_assert
+                        and (
+                            config.test_configs.runtime_triton_dtype_assert
+                            or config.test_configs.static_cpp_dtype_assert
+                        )
                         and dtype is not None
+                        and get_current_backend() != "cpp"
                     ):
-                        assert_line = f"tl.static_assert({self.prefix}{var}.dtype == {triton_type(dtype)})"
-                        buffer.writeline(assert_line)
+                        check_dtype(buffer, var, dtype)
```

Comment:
> Looks good, one comment. Also, there is another place we codegen dtype asserts you should add to cpp as well:
> 
> https://github.com/pytorch/pytorch/pull/146067/files#diff-ecbd0bdf6ee8fc4eab27af1ec0b4369962dd340547451ac2f396b8f12f2f41bcR1788-R1796

It's here. However, I found that CPP codegen usually generate a variable and set `is_vec` later (example below), which makes type checking hard and verbose if we want to check both vectorized and scalar type. So, I decide to leave it for now until we find a way to set is_vec in `generate`. 
https://github.com/pytorch/pytorch/blob/93316cfe945117169db802c0d1886892041bf0d3/torch/_inductor/codegen/cpp.py#L2961-L2967